eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/mutable/AbstractHashBag.java

```diff
@@ -67,8 +67,7 @@ public boolean equals(Object other)
         {
             return false;
         }
-
-        return this.items.keyValuesView().allSatisfy(each -> bag.occurrencesOf(each.getOne()) == each.getTwo());
+        return this.items.shortCircuitKeyValue((k, v) -> bag.occurrencesOf(k) == v, false, false, true);
```

Comment:
Did you performance test an implementation like this?

```java
return this.items.keysView().allSatisfy(each -> bag.occurrencesOf(each) == this.items.get(each));
```

I could see it being slower because the extra `this.items.get` might add more time than creating the Pair wrappers wastes, but I'm curious.

Reply:
If we implemented lazy map views, the implementation here would become:

```java
return this.items.asLazyMap().allSatisfy((key, value) -> bag.occurrencesOf(key) == value);
```

Reply:
Did you performance test an implementation like this?

```java

        if (bag instanceof AbstractHashBag)
        {
            AbstractHashBag<?> that = (AbstractHashBag<?>) other;
            return this.items.equals(that.items);
        }

        return bag.equals(this);
```

Reply:
> Did you performance test an implementation like this?
> 
> ```java
> return this.items.keysView().allSatisfy(each -> bag.occurrencesOf(each) == this.items.get(each));
> ```
> 
> I could see it being slower because the extra `this.items.get` might add more time than creating the Pair wrappers wastes, but I'm curious.

Yes, it was significantly slower.

Reply:
> Did you performance test an implementation like this?
> 
> ```java
>         if (bag instanceof AbstractHashBag)
>         {
>             AbstractHashBag<?> that = (AbstractHashBag<?>) other;
>             return this.items.equals(that.items);
>         }
> 
>         return bag.equals(this);
> ```

Yes, this was slower than the current implementation, most likely because it has to do `containsKey` and they `getOrThrow`.